### ISSUE TITLE
Multi-keystroke patterns can't be created solely out of modifier keys. (like 'shift shift')

### ISSUE THREAD

**Issue by krilnon** [atom/atom-keymap#21]
It's hard to tell if this is by design or not, but I wanted to make a _double tap_ of the shift key open the autocomplete menu.  This approach works for doubled special keys and character literals, so I expected it to work for modifiers too.

Doesn't work: 

``` coffeescript
'.editor':
  'shift shift': 'command-palette:toggle'
```

---

Does work:

``` coffeescript
'.editor':
  'escape escape': 'command-palette:toggle'
```

``` coffeescript
'.editor':
  'x x': 'command-palette:toggle'
```

(I toggled the command palette in these examples but the same applies for autocomplete, etc.)

Is this broken or by design?  And if it's by design, why?  

Thanks! :hamster:

**Comment by bramcordie** [comment]
Is it possible to configure around this design limitation? I'm used to double tapping ⇧ to launch *Search Everywhere* in PhpStorm. I would like atom to behave the same way with the following binding:

```cson
'.editor':
  'shift shift': 'fuzzy-finder:toggle-file-finder'
```

**Comment by blmarket** [comment]
Also IntelliJ(seems most of JetBrains IDEs) also supports :arrow_up: :arrow_up: to 'Search Everywhere'

**Comment by NV0** [comment]
+1

**Comment by cakebake** [comment]
+1

**Comment by blackfyre** [comment]
:+1:

**Comment by shaunhAU** [comment]
+1

**Comment by ruleb** [comment]
+1

**Comment by remcohh** [comment]
+1

**Comment by nunobaba** [comment]
+1
Would it imply a lot of changes to implement this?

**Comment by nathansobo** [comment]
> Would it imply a lot of changes to implement this?

The big problem is distinguishing intentional bare-modifier keystrokes from someone just hitting the modifier key immediately prior to another character... for example, we need to distinguish the sequence `ctrl-a cmd-p` from `ctrl ctrl-a cmd cmd-p`.

Recognizing the bare modifiers as their own keystroke beyond the first in the sequence creates ambiguity in this regard. It's potentially solvable by matching against both the sequence including bare modifiers and the sequence with the modifiers stripped, but doing so does involve enough changes that I can't take the time to dive in and figure it out right now.

**Comment by jnmandal** [comment]
+1 

I have to context switch between inteli-j and atom for work and I would love to add 'shift shift' into Atom.

**Comment by nunobaba** [comment]
@nathansobo Do you have any idea which part of the code is responsible for the keystroke detection? I might have the time to dive into it.

**Comment by nathansobo** [comment]
@nunobaba [This](https://github.com/atom/atom-keymap/blob/master/src/keymap-manager.coffee#L400) is the line where we prevent handling of bare modifiers. You can follow the logic from there. It's going to take some creativity to allow multi-stroke bare modifier matches without interfering with ordinary matches. We currently just compare strings, for example, [here](https://github.com/atom/atom-keymap/blob/master/src/keymap-manager.coffee#L529). We could potentially perform a more sophisticated comparison where bare modifiers are selectively ignored.

**Comment by agjs** [comment]
+ for this feature, would really make Atom even more amazing or to invent a new word, Amazingier. Hope to see it in the upcoming version.

**Comment by harryPL** [comment]
+1. I use the 'shift shift' in RubyMine all the time and it makes jumping between files really peachy.

**Comment by vtambourine** [comment]
Count me it! I am also long to use this WebStorm feature in Atom!

**Comment by SavePointSam** [comment]
+1 for implementation. Switch between Webstorm and Atom daily. Would like the single 'shift shit' key command for both. :D

**Comment by chrisfilo** [comment]
+1

**Comment by andrewhoff** [comment]
+1 this would be awesome!

**Comment by mikebronner** [comment]
looking for this as well. :) here's i what i'm using: 

```
'.platform-darwin':
  'ctrl-shift': 'fuzzy-finder:toggle-file-finder
```

**Comment by MikeJKBird** [comment]
+1

**Comment by KrissiK** [comment]
+1

**Comment by Xenostar** [comment]
+1

**Comment by AndersonJo** [comment]
Please add something like this.. (this one is used in Pycharm)
'.platform-linux':
  'shift shift': 'fuzzy-finder:toggle-file-finder'

**Comment by AndersonJo** [comment]
+1

**Comment by CCristi** [comment]
+1

**Comment by WunderEye** [comment]
+1

**Comment by JBird608** [comment]
+1

**Comment by MovingGifts** [comment]
Any progress on this?

**Comment by VladShcherbin** [comment]
Oh, come on guys, two years since issue was created. We need this, is it that hard to implement?

**Comment by 50Wliu** [comment]
@VladShcherbin: Yes it is, please see @nathansobo's comment [above](https://github.com/atom/atom-keymap/issues/21#issuecomment-145252836).

**Comment by spurdow** [comment]
*bump* +1

**Comment by lordblendi** [comment]
_bump_ +1 pls

**Comment by coxjc** [comment]
Begging for this...

**Comment by 50Wliu** [comment]
Thanks for the feedback everyone! However, all these +1s and _bumps_ are just adding noise, so to minimize that I'm going to lock this issue for now – we may unlock it once someone starts working on this enhancement to get further feedback.
